---
title: "fastQC reports"
author: "SÃ©bastien Renaut (sebastien.renaut@criucpq.ulaval.ca)"
date: "`r Sys.time()`"
output:
  html_document:
    number_sections: T
params:
  fq.dir: 'C:/Users/renseb01/Documents/rnaseq/data/fastq2'
  qc.dir: 'C:/Users/renseb01/Documents/rnaseq/data/fastqc_results'
  threads: 12
  fastqc.path: 'C:/Program\ Files/FastQC/fastqc' 
  metadata: 'C:/Users/renseb01/Documents/rnaseq/data/librairies_1_a_70_et_RIN.xlsx'
  adapters.dir: NULL
---


# Introduction  
* This is a fastQC report of raw sequences. I uses functions from `fastqcr`, which is itself a wrapper for [`fastqc`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
* You can knit this directly from `Rstudio` or you can create this document by running `Rscript -e "rmarkdown::render('fastqc_reports.Rmd',params = list())"` on a Linux/Unix machine. 


 
 
```{r, timehook,echo=F}
knitr::knit_hooks$set(timehook = local({
  now <- NULL
  function(before, options) {
    if (before) {
      # record the current time before each chunk
      now <<- Sys.time()
    } else {
      # calculate the time difference after a chunk
      res <- difftime(Sys.time(), now)
      # return a character string to show the time
      paste("\n\n\nTime for this code chunk to run:",signif(res,3),attr(res,'units'))
    }
  }
}))
``` 

```{r, requirements,echo=TRUE,message =FALSE,warning =F}
library(fastqcr)
library(DT)
library(readxl)
library(dplyr)
library(ggplot2)
library(tidyr)
library(plotly)
```

```{r, functions,echo = F}
###hack for windows
fastqc_windows = function (fq.dir = getwd(), qc.dir = NULL, threads = 8, fastqc.path = "", adapters.dir = NULL) 
{
  fq.dir = sub('C:','/mnt/c',fq.dir)
  qc.dir = sub('C:','/mnt/c',qc.dir)
  adapters.dir = sub('C:','/mnt/c',adapters.dir)
  fastqc.path = sub('C:','wsl.exe /mnt/c',fastqc.path)

  if (is.null(qc.dir)) 
    qc.dir <- file.path(fq.dir, "FASTQC")
  #.create_dir(qc.dir)
  cmd <- paste0(fastqc.path, " ", fq.dir, "/*R1.fastq.gz ",fq.dir,"/*R2.fastq.gz --threads ", 
                threads, " --outdir ", qc.dir, ifelse(is.null(adapters.dir),"",paste0(' --adapters ',adapters.dir)))
  system(cmd)
}

####
fastqc2 = function (fq.dir = getwd(), qc.dir = NULL, threads = 8, fastqc.path = "", adapters.dir = NULL) 
{
    if (is.null(qc.dir)) 
        qc.dir <- file.path(fq.dir, "FASTQC")
    #.create_dir(qc.dir)
    cmd <- paste0(fastqc.path, " ", fq.dir, "/*R1.fastq.gz ",fq.dir,"/*R2.fastq.gz --threads ", 
        threads, " --outdir ", qc.dir, ifelse(is.null(adapters.dir),"",paste0(' --adapters ',adapters.dir)))
    system(cmd)
}





fastqcr_wrapper = function(fq.dir = 'path',
                           qc.dir = 'path', 
                           threads = 8,
                           fastqc.path = 'path',
                           metadata.dir = NULL,
                           adapters.dir = NULL){
 
 if(length(list.files(qc.dir))==0) {
   if(Sys.info()['sysname'] != 'Windows') {
    fastqc2(qc.dir = qc.dir ,fq.dir = fq.dir,threads = threads,fastqc.path = fastqc.path, adapters.dir = adapters.dir)}
  
    if(Sys.info()['sysname'] == 'Windows') {
    message('waning: Running a modified version of fastqc() for Windows');
    fastqc_windows(qc.dir = qc.dir,
                   fq.dir = fq.dir,
                   threads = threads,
                   fastqc.path = fastqc.path,
                   adapters.dir = adapters.dir)}
 } else {message(paste0('Directory ',qc.dir, 'is  not empty. Will not run fastqc and assume that it has already run.'))}
  
  #zip the whole thing
  zip_cmd = paste0('zip -j ',file.path(qc.dir,"../fastqc_individual_reports.zip "),file.path(qc.dir,"*fastqc.zip"))
  system(zip_cmd)
  
  message(paste0(Sys.time(),
                 '--- Zip archive is stored here: ',
                 file.path(fq.dir,"../fastqc_individual_reports.zip")))

  #aggregate results
  qc = suppressMessages(qc_aggregate(qc.dir = qc.dir, progressbar = F))
  
  
  # Generates a summary of qc_aggregate
  qc_summary = summary(qc)
  
  # General statistics of fastqc reports.
  qc_stats = qc_stats(qc)
  qc_stats$ends = sapply(strsplit(qc_stats$sample,split = '_'),tail, 1)
  qc_stats$bio_sample = gsub('_R[12]','',qc_stats$sample)
  
  
  #fix (shorten) names
  if(length(grep('i5.',qc_stats$bio_sample[1]))==1) {
    qc_stats$bio_sample = sapply(strsplit(qc_stats$bio_sample,split ='i5.',fixed = T),"[[", 2)
  }
  
  #qc_read_collection
  qc.files <- list.files(qc.dir,pattern = '.zip', full.names = TRUE)
  qc.names = gsub("_fastqc.zip","",list.files(qc.dir,pattern = '.zip',full.names = F))

  qc_read_collection <- suppressMessages(qc_read_collection(qc.files,sample_names = qc.names,modules = c("all")))
  
  mean_quality_perfile = qc_read_collection$per_base_sequence_quality %>% select(c('sample','Mean')) %>% group_by(sample) %>% summarise(mean_quality= mean(Mean))
  
  mean_quality_perfile$mean_quality = signif(mean_quality_perfile$mean_quality,4)
  
  qc_stats = merge(qc_stats,mean_quality_perfile)
  
  # metadata
  if(!is.null(metadata.dir) & file.exists(metadata.dir)){
    metadata = read_xlsx(metadata.dir)
    rin = metadata %>% select(c(Nom...1,RIN))
    colnames(rin)[1] = 'bio_sample'
    qc_stats =  merge(qc_stats,rin)
  }
  
  return(list(qc_stats,qc_summary,qc_read_collection))
}
```

# FastQC parameters
```{r, parameters,message = T, echo =T}
print(params)
```

# Running FastQC functions
```{r, tables,message = T, eval = T, echo = T, timehook = T}
fastqc_statistics= fastqcr_wrapper(fq.dir = params$fq.dir,
                qc.dir = params$qc.dir,
                threads = params$threads,
                fastqc.path = params$fastqc.path,
                metadata = params$metadata,
                adapters.dir = params$adapters.dir)
```

# Summary table
  * Listing statistics for each .fastq file
```{r, summary_tables,echo = F}
  
  #datatable 
  datatable(fastqc_statistics[[1]],caption = htmltools::tags$caption(htmltools::strong("Summary statistics"), style="color:darkred"))
``` 


# Summary table (detailed)
  * Listing summary statistics for the whole cohort
```{r, summary_detailed,echo = F}
  #datatable
  datatable(fastqc_statistics[[2]],caption = htmltools::tags$caption(htmltools::strong("Summary statistics (detailed)"), style="color:darkred"))
```


# QC Plots
 * A few plots that might be worth looking at (but remember that we have all individual plots).
```{r, plots,timehook =T,fig.width = 11, echo=T}
# Plot per sequence GC content
qc_read_collection = fastqc_statistics[[3]]
ggplotly(qc_plot_collection(qc_read_collection, "Per sequence GC content"))

# Per base sequence quality
ggplotly(qc_plot_collection(qc_read_collection, "Per base sequence quality"))

# Per sequence quality scores
ggplotly(qc_plot_collection(qc_read_collection, "Per sequence quality scores"))

# Overrepresented sequences
ggplotly(qc_plot_collection(qc_read_collection, "Overrepresented sequences"))

# Adapter content
if(is.numeric(qc_read_collection$adapter_content$Position)==F) {
newmean_position = c(1:length(qc_read_collection$adapter_content$Position))[regexpr('-',qc_read_collection$adapter_content$Position,fixed = T)>0]
means = strsplit(qc_read_collection$adapter_content$Position[newmean_position],'-')
means = sapply(means,function(x) mean(as.numeric(x)))
qc_read_collection$adapter_content$new_position = ''
qc_read_collection$adapter_content$new_position[newmean_position] = means
qc_read_collection$adapter_content$new_position[-newmean_position] = qc_read_collection$adapter_content$Position[-newmean_position]
qc_read_collection$adapter_content$new_position = as.numeric(qc_read_collection$adapter_content$new_position)
}
if(is.numeric(qc_read_collection$adapter_content$Position)==T) {
  qc_read_collection$adapter_content$new_position = qc_read_collection$adapter_content$Position
}

adapter_data = qc_read_collection$adapter_content[,-2] %>% pivot_longer(cols = !c(sample,new_position))

adapter_plot = adapter_data %>%
  ggplot( aes(x=new_position, y=value, group=sample, color=sample)) +
    geom_line() +
    ylim(0, 100) + ylab('percentage') + ggtitle('adapter (all) content')

ggplotly(adapter_plot)

```